File: app/backend/app/schemas/recurring_expense.py
# app/backend/app/schemas/recurring_expense.py
from __future__ import annotations

from datetime import date, datetime
from typing import Literal

from pydantic import BaseModel, Field

IntervalUnit = Literal["days", "weeks", "months"]


class RecurringExpenseCreate(BaseModel):
    city_id: str | None = None
    name: str = Field(min_length=1, max_length=120)
    category: str = Field(min_length=1, max_length=80)
    amount_cents: int = Field(gt=0)
    payment_method: str | None = Field(default=None, max_length=80)
    vendor_customer: str | None = Field(default=None, max_length=120)
    description: str | None = None
    interval_value: int = Field(gt=0)
    interval_unit: IntervalUnit
    next_due_date: date


class RecurringExpensePatch(BaseModel):
    name: str | None = Field(default=None, min_length=1, max_length=120)
    category: str | None = Field(default=None, min_length=1, max_length=80)
    amount_cents: int | None = Field(default=None, gt=0)
    payment_method: str | None = None
    vendor_customer: str | None = None
    description: str | None = None
    interval_value: int | None = Field(default=None, gt=0)
    interval_unit: IntervalUnit | None = None
    next_due_date: date | None = None
    is_active: bool | None = None


class RecurringExpenseOut(BaseModel):
    id: str
    city_id: str
    name: str
    category: str
    amount_cents: int
    payment_method: str | None
    vendor_customer: str | None
    description: str | None
    interval_value: int
    interval_unit: IntervalUnit
    next_due_date: date
    google_calendar_event_id: str | None
    is_active: bool
    created_by: str | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: app/frontend/src/types/recurringExpense.ts
// app/frontend/src/types/recurringExpense.ts

export type IntervalUnit = 'days' | 'weeks' | 'months'

export interface RecurringExpense {
  id: string
  city_id: string
  name: string
  category: string
  amount_cents: number
  payment_method: string | null
  vendor_customer: string | null
  description: string | null
  interval_value: number
  interval_unit: IntervalUnit
  next_due_date: string  // ISO date string YYYY-MM-DD
  google_calendar_event_id: string | null
  is_active: boolean
  created_by: string | null
  created_at: string
  updated_at: string
}

export interface RecurringExpenseInput {
  city_id?: string | null
  name: string
  category: string
  amount_cents: number
  payment_method?: string | null
  vendor_customer?: string | null
  description?: string | null
  interval_value: number
  interval_unit: IntervalUnit
  next_due_date: string
}

export interface RecurringExpensePatch {
  name?: string
  category?: string
  amount_cents?: number
  payment_method?: string | null
  vendor_customer?: string | null
  description?: string | null
  interval_value?: number
  interval_unit?: IntervalUnit
  next_due_date?: string
  is_active?: boolean
}

export interface LogResult {
  transaction_id: string
  next_due_date: string
  gcal_warning: string | null
}

export function intervalLabel(value: number, unit: IntervalUnit): string {
  if (value === 1) {
    const singular: Record<IntervalUnit, string> = { days: 'day', weeks: 'week', months: 'month' }
    return `Every ${singular[unit]}`
  }
  return `Every ${value} ${unit}`
}

export function centsToDisplay(cents: number): string {
  return new Intl.NumberFormat('en-US', { style: 'currency', currency: 'USD' }).format(cents / 100)
}

export function isOverdue(nextDueDate: string): boolean {
  return nextDueDate < new Date().toLocaleDateString('en-CA')
}
